substrate-archive/src/actors/workers/extrinsics_decoder.rs

```diff
@@ -0,0 +1,139 @@
+// Copyright 2017-2021 Parity Technologies (UK) Ltd.
+// This file is part of substrate-archive.
+
+// substrate-archive is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+// substrate-archive is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with substrate-archive.  If not, see <http://www.gnu.org/licenses/>.
+
+use arc_swap::ArcSwap;
+use async_std::task;
+use sqlx::PgPool;
+use itertools::Itertools;
+use std::{collections::HashMap, sync::Arc};
+use xtra::prelude::*;
+
+use desub::Decoder;
+
+use crate::{
+	actors::{
+		workers::database::{DatabaseActor, GetState},
+		SystemConfig,
+	},
+	database::{models::ExtrinsicsModel, queries},
+	error::{ArchiveError, Result},
+	types::BatchExtrinsics,
+};
+
+pub struct ExtrinsicsDecoder {
+	pool: PgPool,
+	addr: Address<DatabaseActor>,
+	max_block_load: u32,
+	decoder: Arc<Decoder>,
+	/// Hashmap of spec version -> number of block that upgrades to that spec version
+	upgrades: ArcSwap<HashMap<u32, u32>>,
+}
+
+impl ExtrinsicsDecoder {
+	pub async fn new<B: Send + Sync, Db: Send + Sync>(
+		config: &SystemConfig<B, Db>,
+		addr: Address<DatabaseActor>,
+	) -> Result<Self> {
+		let max_block_load = config.control.max_block_load;
+		let chain = config.persistent_config.chain();
+		let pool = addr.send(GetState::Pool).await??.pool();
+		let decoder = Arc::new(Decoder::new(chain));
+		let mut conn = pool.acquire().await?;
+		let upgrades = ArcSwap::from_pointee(queries::upgrade_blocks_from_spec(&mut conn, 0).await?);
+		log::info!("Started extrinsic decoder");
+		Ok(Self { pool, addr, max_block_load, decoder, upgrades })
+	}
+
+	async fn crawl_missing_extrinsics(&mut self) -> Result<()> {
+		let mut conn = self.pool.acquire().await?;
+		let blocks = queries::missing_extrinsic_blocks(&mut conn, self.max_block_load).await?;
+		log::info!("Indexing {} missing extrinsic blocks", blocks.len());
+
+		let versions: Vec<u32> =
+			blocks.iter().filter(|b| !self.decoder.has_version(&b.3)).map(|(_, _, _, v)| *v).unique().collect();
+		// above and below line are separate to let immutable ref to `self.decoder` to go out of scope.
+		for version in versions.iter() {
+			let metadata = queries::metadata(&mut conn, *version as i32).await?;
+			log::debug!("Registering version {}", version);
+			// TODO: FIX EXPECT
+			Arc::get_mut(&mut self.decoder)
+				.expect("Actors guarantee one reference; qed")
```

Comment:
This doesn't necessarily hold true, i should make expect into error or remove arc and use a different solution